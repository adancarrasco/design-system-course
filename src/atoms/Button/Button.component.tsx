import React from 'react'
import styled from '@emotion/styled'

export interface IButtonProps {
  StartIcon?: React.ReactType
  children?: React.ReactNode | string
  EndIcon?: React.ReactType
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void
}

const resolveChildren = (children) =>
  typeof children === 'function' ? children : <span>{children}</span>

const Button: React.FC<IButtonProps> = (
  { StartIcon, children, EndIcon, onClick = () => {} },
  ...rest
) => {
  return (
    <ButtonStyled onClick={onClick} {...rest}>
      {StartIcon && <StartIcon />}
      {children && resolveChildren(children)}
      {EndIcon && <EndIcon />}
    </ButtonStyled>
  )
}

interface ButtonStyledProps {
  color?: string
  backgroundColor?: string
  backgroundColorHover?: string
  fontSize?: string
  padding?: string
}

const ButtonStyled = styled.button<ButtonStyledProps>`
  color: ${(props) => props.color ?? '#fff'};
  background-color: ${(props) => props.backgroundColor ?? 'rgba(130, 130, 130, 0.5)'};
  font-size: ${(props) => props.fontSize ?? '1rem'};
  padding: ${(props) => props.padding ?? '0.5rem'};
  cursor: pointer;
  border: none;
  border-radius: 4px;
  & svg:first-of-type {
    margin-right: 4px;
  }
  & svg:last-child {
    margin-left: 4px;
  }
  & svg:only-child {
    margin: 0;
  }
  & > svg,
  & > span {
    vertical-align: middle;
  }
  &:hover {
    background-color: ${(props) => props.backgroundColorHover ?? 'rgba(145, 145, 145, 0.5)'};
  }
`

export default Button
