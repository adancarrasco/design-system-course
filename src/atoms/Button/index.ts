export { default } from './Button.component'
export type { IButtonProps } from './Button.component'
