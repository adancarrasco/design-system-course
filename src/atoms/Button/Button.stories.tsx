import React from 'react'
import { storiesOf } from '@storybook/react'
import { AccessAlarm, ThreeDRotation } from '@material-ui/icons'

import Button from '.'

storiesOf('Atoms|Button', module)
  .add('Default', () => <Button>This is a button</Button>)
  .add('With StartIcon', () => <Button StartIcon={AccessAlarm}>Access Alarm Icon</Button>)
  .add('With EndIcon', () => <Button EndIcon={ThreeDRotation}>ThreeDRotation</Button>)
  .add('With both icons', () => (
    <Button StartIcon={AccessAlarm} EndIcon={ThreeDRotation}>
      Both Icons
    </Button>
  ))
  .add('One icon', () => <Button StartIcon={AccessAlarm} />)
