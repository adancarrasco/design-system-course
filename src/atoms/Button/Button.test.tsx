import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Button, { IButtonProps } from '.'

describe('Button Component', () => {
  const renderComponent = ({ children }: IButtonProps) => render(<Button>{children}</Button>)
  it('matches snapshot', () => {
    const props = {
      children: 'Hello',
    }
    const { getByText } = renderComponent(props)
    const container = getByText('Hello')
    expect(container).toMatchSnapshot()
  })

  it('should render', () => {
    const props = {
      children: 'Hello',
    }
    const { getByText } = renderComponent(props)
    const container = getByText('Hello')
    expect(container).toBeInTheDocument()
  })

  it('should render with startIcon', () => {
    const props = {
      children: 'Hello',
      icon: 'myIcon',
    }
    const { getByText } = renderComponent(props)
    const container = getByText('Hello')
    expect(container.previousSibling).toBe('myIcon')
  })
})
