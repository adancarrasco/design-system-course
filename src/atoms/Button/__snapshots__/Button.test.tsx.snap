// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
<span>
  Hello
</span>
`;

exports[`Button Component matches snapshot with StartIcon 1`] = `
<span>
  Hello
</span>
`;
